File: abc/go/abc210/abc210_E/abc210_E.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)
var wrtr = bufio.NewWriterSize(os.Stdout, 10_000_000)
var rdr = bufio.NewScanner(os.Stdin)
func gs() string  { rdr.Scan(); return rdr.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gi2() (int,int) { return gi(),gi() }
func ia(m int) []int { return make([]int,m) }
func fill2(m int) ([]int,[]int) { a,b := ia(m),ia(m); for i:=0;i<m;i++ {a[i],b[i] = gi(),gi()}; return a,b }
func gcd(a,b int) int { for b != 0 { t:=b; b=a%b; a=t }; return a }
func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	defer wrtr.Flush()
	infn := ""; if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = bufio.NewScanner(f) }
	rdr.Split(bufio.ScanWords); rdr.Buffer(make([]byte,1024),1_000_000_000)
	// PROGRAM STARTS HERE
	type op struct {a,c int}
	N,M := gi2(); A,C := fill2(M)
	ops := make([]op,M)
	for i:=0;i<M;i++ { ops[i] = op{A[i],C[i]} }
	sort.Slice(ops,func(i,j int)bool{return ops[i].c < ops[j].c})
	ans := 0
	for _,xx := range ops {
		inc := xx.a % N
		if inc == 0 { continue }
		g := gcd(inc,N)
		ans,N = ans+xx.c*(N-g),g
	}
	if N == 1 { fmt.Println(ans) } else { fmt.Println(-1) }
}

File: abc/python/abc210/abc210_E.py

import sys
import math
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N,M = gis()
    A = [0] * M
    C = [0] * M
    for i in range(M) :A[i],C[i] = gis()
    edges = [c << 30 | a for (a,c) in zip(A,C)]
    edges.sort()
    cost = 0
    for x in edges :
        (c,a) = (x >> 30, x & 0x3fffffff)
        g = math.gcd(N,a)
        if g == N : continue
        cost += (N-g) * c
        N = g
    ans = -1 if N > 1 else cost
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

